src/cargo/util/config/de.rs

```diff
@@ -0,0 +1,447 @@
+    {
+        match self.next.take().expect("next field missing") {
+            KeyKind::Normal(key) => self.de.key.push(&key),
+            KeyKind::CaseSensitive(key) => self.de.key.push_sensitive(&key),
+        }
+        let result = seed.deserialize(Deserializer {
+            config: self.de.config,
+            key: self.de.key.clone(),
+        });
+        self.de.key.pop();
+        return result;
+    }
+}
+
+struct ConfigSeqAccess {
+    list_iter: vec::IntoIter<(String, Definition)>,
+}
+
+impl ConfigSeqAccess {
+    fn new(de: Deserializer<'_>) -> Result<ConfigSeqAccess, ConfigError> {
+        let mut res = Vec::new();
+        if let Some(v) = de.config.get_list(de.key.as_config_key())? {
+            for (s, path) in v.val {
+                res.push((s, Definition::Path(path)));
+            }
+        }
+
+        if de.config.cli_unstable().advanced_env {
+            // Parse an environment string as a TOML array.
+            if let Some(v) = de.config.env.get(de.key.as_env_key()) {
+                let def = Definition::Environment(de.key.as_env_key().to_string());
+                if !(v.starts_with('[') && v.ends_with(']')) {
+                    return Err(ConfigError::new(
+                        format!("should have TOML list syntax, found `{}`", v),
+                        def,
+                    ));
+                }
+                let toml_s = format!("value={}", v);
+                let toml_v: toml::Value = toml::de::from_str(&toml_s).map_err(|e| {
+                    ConfigError::new(format!("could not parse TOML list: {}", e), def.clone())
+                })?;
+                let values = toml_v
+                    .as_table()
+                    .unwrap()
+                    .get("value")
+                    .unwrap()
+                    .as_array()
+                    .expect("env var was not array");
+                for value in values {
+                    // TODO: support other types.
+                    let s = value.as_str().ok_or_else(|| {
+                        ConfigError::new(
+                            format!("expected string, found {}", value.type_str()),
+                            def.clone(),
+                        )
+                    })?;
+                    res.push((s.to_string(), def.clone()));
+                }
+            }
+        }
+        Ok(ConfigSeqAccess {
+            list_iter: res.into_iter(),
+        })
+    }
+}
+
+impl<'de> de::SeqAccess<'de> for ConfigSeqAccess {
+    type Error = ConfigError;
+
+    fn next_element_seed<T>(&mut self, seed: T) -> Result<Option<T::Value>, Self::Error>
+    where
+        T: de::DeserializeSeed<'de>,
+    {
+        match self.list_iter.next() {
+            // TODO: add `def` to error?
+            Some((value, _def)) => seed.deserialize(value.into_deserializer()).map(Some),
+            None => Ok(None),
+        }
+    }
+}
+
+struct ValueDeserializer<'config> {
+    hits: u32,
```

Comment:
Can there be (here or maybe in `config/value.rs`), a high-level overview of the strategy for how these magic fields are begin used?   I think I sorta understand it, but I think an overview would go a long way to help.